modules/programs/carapace.nix

```diff
@@ -0,0 +1,88 @@
+{ config, pkgs, lib, ... }:
+
+let
+  inherit (lib) mkEnableOption mkPackageOption mkIf pipe fileContents splitString;
+  cfg = config.programs.carapace;
+  bin = cfg.package + "/bin/carapace";
+in {
+  meta.maintainers = with lib.maintainers; [ weathercold bobvanderlinden ];
+
+  options.programs.carapace = {
+    enable =
+      mkEnableOption "carapace, a multi-shell multi-command argument completer";
+    package = mkPackageOption pkgs "carapace" { };
+    enableBashIntegration = mkEnableOption "Bash integration" // {
+      default = true;
+    };
+    enableZshIntegration = mkEnableOption "Zsh integration" // {
+      default = true;
+    };
+    enableFishIntegration = mkEnableOption "Fish integration" // {
+      default = true;
+    };
+    enableNushellIntegration = mkEnableOption "Nushell integration" // {
+      default = true;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+
+    programs = {
+      bash.initExtra = mkIf cfg.enableBashIntegration ''
+        source <(${bin} _carapace bash)
+      '';
+
+      zsh.initExtra = mkIf cfg.enableZshIntegration ''
+        source <(${bin} _carapace zsh)
+      '';
+
+      fish.interactiveShellInit = mkIf cfg.enableFishIntegration ''
+        ${bin} _carapace fish | source
+      '';
+
+      nushell = mkIf cfg.enableNushellIntegration {
+        extraEnv = ''
+          let carapace_cache = "${config.xdg.cacheHome}/carapace"
+          if not ($carapace_cache | path exists) {
+            mkdir $carapace_cache
+          }
+          ${bin} _carapace nushell | save -f $"($carapace_cache)/init.nu"
+        '';
+        extraConfig = ''
+          source ${config.xdg.cacheHome}/carapace/init.nu
+        '';
+      };
+    };
+
+    home.file = mkIf (config.programs.fish.enable && cfg.enableFishIntegration) (
```

Comment:
put it in `config`, otherwise it will be enabled for all fish users

use `xdg.configFile` instead and remove `.config/` below

Reply:
> put it in `config`, otherwise it will be enabled for all fish users

I'm not exactly sure what you mean by this? Both `config.programs.fish.enable` and `cfg.enableFishIntegration` need to be enabled for this to be the case?

> use `xdg.configFile` instead and remove `.config/` below

Good suggestion :+1:. Updated the PR.

Reply:
`cfg.enableFishIntegration` is enabled by default, so these empty files will be generated just by enabling fish. You need to put `xdg.configFile` in the `config` attrset above which also checks for `cfg.enable`.

Also, there's one mention of `home.file` left in the comments

Reply:
> `cfg.enableFishIntegration` is enabled by default, so these empty files will be generated just by enabling fish. You need to put `xdg.configFile` in the `config` attrset above which also checks for `cfg.enable`.

https://github.com/bobvanderlinden/home-manager/blob/361cf75cc6eb55c0e215927b313401631727c440/modules/programs/carapace.nix#L79

is already inside

https://github.com/bobvanderlinden/home-manager/blob/361cf75cc6eb55c0e215927b313401631727c440/modules/programs/carapace.nix#L44

I'm probably misunderstanding :sweat_smile:, can you make a code suggestion to explain?

> Also, there's one mention of `home.file` left in the comments

Good catch :+1: I've updated the PR.

Reply:
It's all good now 👍